File: backend/app/schemas/english_speaking.py
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class DifficultyTier(str, Enum):
    LEVEL_1 = "level_1"  # 2-3 word phrases
    LEVEL_2 = "level_2"  # 4-5 word simple sentences
    LEVEL_3 = "level_3"  # Interactive conversational responses


class ErrorCategory(str, Enum):
    PRONUNCIATION = "pronunciation"
    MISSING_WORD = "missing_word"
    EXTRA_WORD = "extra_word"
    FLUENCY_PAUSE = "fluency_pause"
    NONE = "none"


class ChildFeedbackTag(str, Enum):
    STAR_PERFORMER = "star_performer"  # Excellent attempt (90-100%)
    GREAT_EFFORT = "great_effort"  # Good attempt with minor soft errors (70-89%)
    KEEP_TRYING = "keep_trying"  # Needs practice (<70%), gentle encouraging feedback
    SUPER_CLEAR = "super_clear"  # Excellent pronunciation tag
    TRY_AGAIN_GENTLY = "try_again_gently"


class TargetPhoneticGuide(BaseModel):
    word: str = Field(..., description="Target English word")
    phonetic_ipa: str = Field(..., description="IPA phonetic representation")
    simple_phonetic: str = Field(
        ..., description="Child-friendly phonetic breakdown (e.g., 'MAHN-go')"
    )
    audio_sample_url: Optional[str] = Field(None, description="Audio reference URL for target word")


class SpeakingPrompt(BaseModel):
    prompt_id: str = Field(..., description="Unique prompt identifier")
    tier: DifficultyTier = Field(..., description="Difficulty tier (level_1, level_2, level_3)")
    target_sentence: str = Field(..., description="Target sentence for student to speak")
    translation_hindi: Optional[str] = Field(None, description="Hindi translation for context")
    translation_marathi: Optional[str] = Field(None, description="Marathi translation for context")
    phonetics: List[TargetPhoneticGuide] = Field(
        default_factory=list, description="Phonetic breakdown per word"
    )
    theme: str = Field(
        default="daily_life", description="Topic/theme ('family', 'animals', 'greetings', etc.)"
    )


class StudentAudioAttempt(BaseModel):
    student_id: str = Field(..., description="Unique student identifier")
    prompt_id: str = Field(..., description="Prompt ID being attempted")
    audio_base64: Optional[str] = Field(None, description="Base64 encoded student audio recording")
    spoken_transcript: Optional[str] = Field(
        None, description="Transcribed text if STT already processed"
    )
    duration_seconds: float = Field(
        default=0.0, ge=0.0, description="Recording duration in seconds"
    )
    audio_format: str = Field(default="wav", description="Audio format ('wav', 'mp3', 'ogg')")


class WordEvaluationDetail(BaseModel):
    target_word: str = Field(..., description="Target expected word")
    spoken_word: Optional[str] = Field(None, description="Actual spoken word recorded")
    score: float = Field(..., ge=0.0, le=100.0, description="Word-level accuracy score (0-100)")
    error_category: ErrorCategory = Field(
        default=ErrorCategory.NONE, description="Type of soft error detected"
    )
    child_tip: Optional[str] = Field(None, description="Child-friendly gentle pronunciation tip")


class AccuracyScore(BaseModel):
    overall_accuracy: float = Field(
        ..., ge=0.0, le=100.0, description="Overall accuracy percentage (0-100)"
    )
    word_accuracy: float = Field(..., ge=0.0, le=100.0, description="Word level accuracy score")
    phonetic_match_percentage: float = Field(
        ..., ge=0.0, le=100.0, description="Phonetic match percentage score"
    )


class FluencyMetrics(BaseModel):
    words_per_minute: float = Field(
        default=0.0, ge=0.0, description="Estimated speaking rate (WPM)"
    )
    pause_count: int = Field(default=0, ge=0, description="Number of hesitant pauses detected")
    fluency_score: float = Field(
        ..., ge=0.0, le=100.0, description="Overall fluency rating (0-100)"
    )


class GentleFeedback(BaseModel):
    encouragement_message: str = Field(
        ..., description="Warm, positive, child-friendly feedback message"
    )
    feedback_tag: ChildFeedbackTag = Field(..., description="Categorized feedback tag")
    stars_earned: int = Field(..., ge=1, le=3, description="Stars awarded (1-3 stars)")
    improvements: List[str] = Field(default_factory=list, description="Gentle tips for improvement")
    praise_points: List[str] = Field(
        default_factory=list, description="Specific things the student did well"
    )


class SpeakingEvaluationRequest(BaseModel):
    student_id: str = Field(..., description="Unique student identifier")
    prompt_id: str = Field(..., description="Prompt ID attempted")
    spoken_transcript: str = Field(..., description="Spoken transcript to evaluate")
    target_sentence: Optional[str] = Field(
        None, description="Target sentence (optional if prompt_id is registered)"
    )
    duration_seconds: float = Field(default=2.5, ge=0.0, description="Audio duration")


class SpeakingEvaluationResponse(BaseModel):
    evaluation_id: str = Field(..., description="Unique evaluation result ID")
    student_id: str = Field(..., description="Student ID")
    prompt_id: str = Field(..., description="Prompt ID")
    target_sentence: str = Field(..., description="Expected target sentence")
    spoken_transcript: str = Field(..., description="Student spoken transcript")
    accuracy: AccuracyScore = Field(..., description="Accuracy breakdown scores")
    fluency: FluencyMetrics = Field(..., description="Fluency assessment metrics")
    word_details: List[WordEvaluationDetail] = Field(
        default_factory=list, description="Word-by-word soft evaluation"
    )
    feedback: GentleFeedback = Field(..., description="Child-friendly encouraging feedback")
    offline_fallback: bool = Field(
        default=False, description="True if local offline evaluation fallback was used"
    )


class PromptsListResponse(BaseModel):
    tier: Optional[DifficultyTier] = Field(None, description="Filtered tier, if specified")
    prompts: List[SpeakingPrompt] = Field(
        ..., description="List of available speaking practice prompts"
    )
    total_count: int = Field(..., description="Total count of prompts returned")

File: mobile/lib/models/english_speaking.dart
/// English Speaking Models for Class 1 Speaking Practice & Soft Assessment
library;

enum DifficultyTier {
  level_1, // 2-3 word phrases
  level_2, // 4-5 word simple sentences
  level_3, // interactive conversational responses
}

enum ErrorCategory {
  pronunciation,
  missingWord,
  extraWord,
  fluencyPause,
  none,
}

enum ChildFeedbackTag {
  starPerformer,
  greatEffort,
  keepTrying,
  superClear,
  tryAgainGently,
}

class TargetPhoneticGuide {
  final String word;
  final String phoneticIpa;
  final String simplePhonetic;
  final String? audioSampleUrl;

  TargetPhoneticGuide({
    required this.word,
    required this.phoneticIpa,
    required this.simplePhonetic,
    this.audioSampleUrl,
  });

  factory TargetPhoneticGuide.fromJson(Map<String, dynamic> json) {
    return TargetPhoneticGuide(
      word: json['word'] as String,
      phoneticIpa: json['phonetic_ipa'] as String,
      simplePhonetic: json['simple_phonetic'] as String,
      audioSampleUrl: json['audio_sample_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'word': word,
        'phonetic_ipa': phoneticIpa,
        'simple_phonetic': simplePhonetic,
        'audio_sample_url': audioSampleUrl,
      };
}

class SpeakingPrompt {
  final String promptId;
  final String tier;
  final String targetSentence;
  final String? translationHindi;
  final String? translationMarathi;
  final List<TargetPhoneticGuide> phonetics;
  final String theme;

  SpeakingPrompt({
    required this.promptId,
    required this.tier,
    required this.targetSentence,
    this.translationHindi,
    this.translationMarathi,
    this.phonetics = const [],
    this.theme = 'daily_life',
  });

  factory SpeakingPrompt.fromJson(Map<String, dynamic> json) {
    return SpeakingPrompt(
      promptId: json['prompt_id'] as String,
      tier: json['tier'] as String,
      targetSentence: json['target_sentence'] as String,
      translationHindi: json['translation_hindi'] as String?,
      translationMarathi: json['translation_marathi'] as String?,
      phonetics: (json['phonetics'] as List<dynamic>?)
              ?.map((e) => TargetPhoneticGuide.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      theme: json['theme'] as String? ?? 'daily_life',
    );
  }

  Map<String, dynamic> toJson() => {
        'prompt_id': promptId,
        'tier': tier,
        'target_sentence': targetSentence,
        'translation_hindi': translationHindi,
        'translation_marathi': translationMarathi,
        'phonetics': phonetics.map((e) => e.toJson()).toList(),
        'theme': theme,
      };
}

class WordEvaluationDetail {
  final String targetWord;
  final String? spokenWord;
  final double score;
  final String errorCategory;
  final String? childTip;

  WordEvaluationDetail({
    required this.targetWord,
    this.spokenWord,
    required this.score,
    required this.errorCategory,
    this.childTip,
  });

  factory WordEvaluationDetail.fromJson(Map<String, dynamic> json) {
    return WordEvaluationDetail(
      targetWord: json['target_word'] as String,
      spokenWord: json['spoken_word'] as String?,
      score: (json['score'] as num).toDouble(),
      errorCategory: json['error_category'] as String,
      childTip: json['child_tip'] as String?,
    );
  }
}

class AccuracyScore {
  final double overallAccuracy;
  final double wordAccuracy;
  final double phoneticMatchPercentage;

  AccuracyScore({
    required this.overallAccuracy,
    required this.wordAccuracy,
    required this.phoneticMatchPercentage,
  });

  factory AccuracyScore.fromJson(Map<String, dynamic> json) {
    return AccuracyScore(
      overallAccuracy: (json['overall_accuracy'] as num).toDouble(),
      wordAccuracy: (json['word_accuracy'] as num).toDouble(),
      phoneticMatchPercentage: (json['phonetic_match_percentage'] as num).toDouble(),
    );
  }
}

class FluencyMetrics {
  final double wordsPerMinute;
  final int pauseCount;
  final double fluencyScore;

  FluencyMetrics({
    required this.wordsPerMinute,
    required this.pauseCount,
    required this.fluencyScore,
  });

  factory FluencyMetrics.fromJson(Map<String, dynamic> json) {
    return FluencyMetrics(
      wordsPerMinute: (json['words_per_minute'] as num).toDouble(),
      pauseCount: json['pause_count'] as int,
      fluencyScore: (json['fluency_score'] as num).toDouble(),
    );
  }
}

class GentleFeedback {
  final String encouragementMessage;
  final String feedbackTag;
  final int starsEarned;
  final List<String> improvements;
  final List<String> praisePoints;

  GentleFeedback({
    required this.encouragementMessage,
    required this.feedbackTag,
    required this.starsEarned,
    this.improvements = const [],
    this.praisePoints = const [],
  });

  factory GentleFeedback.fromJson(Map<String, dynamic> json) {
    return GentleFeedback(
      encouragementMessage: json['encouragement_message'] as String,
      feedbackTag: json['feedback_tag'] as String,
      starsEarned: json['stars_earned'] as int,
      improvements: (json['improvements'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      praisePoints: (json['praise_points'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
    );
  }
}

class SpeakingEvaluationRequest {
  final String studentId;
  final String promptId;
  final String spokenTranscript;
  final String? targetSentence;
  final double durationSeconds;

  SpeakingEvaluationRequest({
    required this.studentId,
    required this.promptId,
    required this.spokenTranscript,
    this.targetSentence,
    this.durationSeconds = 2.5,
  });

  Map<String, dynamic> toJson() => {
        'student_id': studentId,
        'prompt_id': promptId,
        'spoken_transcript': spokenTranscript,
        'target_sentence': targetSentence,
        'duration_seconds': durationSeconds,
      };
}

class SpeakingEvaluationResponse {
  final String evaluationId;
  final String studentId;
  final String promptId;
  final String targetSentence;
  final String spokenTranscript;
  final AccuracyScore accuracy;
  final FluencyMetrics fluency;
  final List<WordEvaluationDetail> wordDetails;
  final GentleFeedback feedback;
  final bool offlineFallback;

  SpeakingEvaluationResponse({
    required this.evaluationId,
    required this.studentId,
    required this.promptId,
    required this.targetSentence,
    required this.spokenTranscript,
    required this.accuracy,
    required this.fluency,
    required this.wordDetails,
    required this.feedback,
    this.offlineFallback = false,
  });

  factory SpeakingEvaluationResponse.fromJson(Map<String, dynamic> json) {
    return SpeakingEvaluationResponse(
      evaluationId: json['evaluation_id'] as String,
      studentId: json['student_id'] as String,
      promptId: json['prompt_id'] as String,
      targetSentence: json['target_sentence'] as String,
      spokenTranscript: json['spoken_transcript'] as String,
      accuracy: AccuracyScore.fromJson(json['accuracy'] as Map<String, dynamic>),
      fluency: FluencyMetrics.fromJson(json['fluency'] as Map<String, dynamic>),
      wordDetails: (json['word_details'] as List<dynamic>?)
              ?.map((e) => WordEvaluationDetail.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      feedback: GentleFeedback.fromJson(json['feedback'] as Map<String, dynamic>),
      offlineFallback: json['offline_fallback'] as bool? ?? false,
    );
  }
}
